src/core/unif.ml

```diff
@@ -162,19 +173,53 @@ let instantiation : ctxt -> meta -> term array -> term ->
         Some (Bindlib.bind_mvar vs (lift u))
   else None
 
+(** Checking type or not during meta instanciation *)
+type type_check = | NoTypeCheck | TypeCheckInstanciation
+let g_type_check = Stdlib.ref TypeCheckInstanciation
+
 (** [instantiate ctx m ts u] check whether, in a problem [m[ts]=u], [m] can be
     instantiated and, if so, instantiate it. *)
-let instantiate : ctxt -> meta -> term array -> term -> bool =
-  fun ctx m ts u ->
+let instantiate : ctxt -> meta -> term array ->
+  term -> constr list -> bool =
+  fun ctx m ts u initial ->
+  let type_meta_app m ts =
```

Comment:
move this to basics.ml:
```
(** [type_app a ts] returns the type of [add_args x ts] where [x] is any term of type [a], if it exists. *) 
let rec type_app : ctxt -> term -> term list -> term option =
      match (Eval.whnf ctx a), ts with
      | Prod(_,b), t :: ts -> type_app (Bindlib.subst b t) ts
      | _, [] -> Some a
      | _, _ -> None
```
Then replace `    let m_app = type_meta_app m (Array.to_list ts) in` by:
```
    let m_app =
      match type_app ctx !(m.meta_type) (Array.to_list ts) with
      | Some a -> a
      | None -> assert false
  in
```
```

Reply:
can't use Eval in Basics....

I have put type_app in Eval.... WDYT ?

Reply:
Ah, right. What about putting it in infer.ml as it is related to type inference?

Reply:
fixed